File: test/devmode.test.ts
/**
 * Unlocking the developer card.
 *
 *   npm run test:dev
 *
 * Two things have to hold: ten deliberate taps get you in, and a slow drip of
 * taps across a week never does. The second is the one that matters — if the
 * counter accumulated across sessions the card would eventually appear on its
 * own, which is exactly what hiding it was meant to prevent.
 */

// readUnlocked/rememberUnlocked touch sessionStorage, which Node has not got.
// A two-line stand-in is enough to assert the round trip and, more usefully,
// that a browser which throws on storage access reports "locked" rather than
// taking the whole settings screen down with it.
const store = new Map<string, string>();
let throws = false;
(globalThis as { sessionStorage?: unknown }).sessionStorage = {
  getItem: (k: string) => {
    if (throws) throw new Error('denied');
    return store.get(k) ?? null;
  },
  setItem: (k: string, v: string) => {
    if (throws) throw new Error('denied');
    store.set(k, v);
  },
  removeItem: (k: string) => {
    if (throws) throw new Error('denied');
    store.delete(k);
  },
};

import {
  NO_TAPS,
  readUnlocked,
  rememberUnlocked,
  tap,
  tapHint,
  tapsLeft,
  TAP_GAP_MS,
  TAPS_TO_UNLOCK,
  unlocked,
  type TapState,
} from '@/lib/devmode';

let failures = 0;

function check(label: string, ok: boolean, detail = '') {
  if (!ok) failures++;
  console.log(`${ok ? 'PASS' : 'FAIL'}  ${label}${detail ? `  — ${detail}` : ''}`);
}

/** Taps 300ms apart, which is about as fast as a thumb goes. */
function run(n: number, from: TapState = NO_TAPS, start = 1_000): TapState {
  let s = from;
  for (let i = 0; i < n; i++) s = tap(s, start + i * 300);
  return s;
}

function main() {
  check('a fresh state is locked', !unlocked(NO_TAPS));
  check('and asks for ten', tapsLeft(NO_TAPS) === TAPS_TO_UNLOCK);

  check('nine taps is not enough', !unlocked(run(9)));
  check('the tenth opens it', unlocked(run(10)));
  check('and so does the eleventh', unlocked(run(11)));

  /* ------------------------------------------------------------- the gap */

  // A pause means a different intention. Without this the counter would carry
  // over from yesterday's visit to the settings screen.
  const paused = tap(run(9), 1_000 + 8 * 300 + TAP_GAP_MS + 1);
  check('a pause restarts the run', paused.count === 1);
  check('and a restarted run is locked', !unlocked(paused));
  check(
    'nine more after the pause still is not enough',
    !unlocked(run(8, paused, paused.last + 300)),
  );
  check('ten from the pause opens it', unlocked(run(9, paused, paused.last + 300)));

  // Exactly at the gap is still one run: the boundary should not punish a
  // steady, unhurried tapper.
  const atTheEdge = tap({ count: 4, last: 1_000 }, 1_000 + TAP_GAP_MS);
  check('a tap exactly on the gap continues the run', atTheEdge.count === 5);

  /* ------------------------------------------------------------- the hint */

  check('nothing is said early on', tapHint(run(3)) === null);
  check('nor at six', tapHint(run(6)) === null);
  check('three left is announced', tapHint(run(7)) === '3 more taps');
  check('and counts down', tapHint(run(8)) === '2 more taps');
  check('the last one is singular', tapHint(run(9)) === '1 more tap');
  check('and it goes quiet once open', tapHint(run(10)) === null);

  /* --------------------------------------------------- staying unlocked */

  /*
    The card used to close every time you left Settings, which is the one
    moment a notification test requires — you have to leave to see whether the
    thing arrived. Ten more taps to get back in is how a test bench stops being
    used.
  */
  check('closed to begin with', !readUnlocked());
  rememberUnlocked(true);
  check('and open once remembered', readUnlocked());
  rememberUnlocked(false);
  check('Hide closes it again', !readUnlocked());

  // A browser that refuses storage must report locked, not throw. Nothing
  // behind the card is worth taking the settings screen down for.
  rememberUnlocked(true);
  throws = true;
  check('storage that throws reads as locked', readUnlocked() === false);
  let survived = true;
  try {
    rememberUnlocked(false);
  } catch {
    survived = false;
  }
  check('and writing to it does not throw', survived);
  throws = false;

  console.log(failures === 0 ? '\nall green' : `\n${failures} failure(s)`);
  process.exit(failures === 0 ? 0 : 1);
}

main();

File: mobile/test/devmode_test.dart
/// Ten taps on the version pill.
///
///   dart test test/devmode_test.dart
///
/// Translated from `test/devmode.test.ts`. The rule with a bug in it is the
/// gap: without it the counter accumulates across a week of ordinary use and
/// the developer card appears unbidden on somebody's settings screen.
library;

import 'package:stash_it/logic/devmode.dart';
import 'package:flutter_test/flutter_test.dart';

final start = DateTime(2026, 8, 12, 9);

DateTime at(int ms) => start.add(Duration(milliseconds: ms));

/// Taps `n` times, 200ms apart — a plausible thumb.
TapState drum(int n) {
  var state = noTaps;
  for (var i = 0; i < n; i++) {
    state = tap(state, at(i * 200));
  }
  return state;
}

void main() {
  group('counting', () {
    test('the first tap starts a run', () {
      expect(tap(noTaps, at(0)).count, 1);
    });

    test('a quick second continues it', () {
      expect(tap(tap(noTaps, at(0)), at(200)).count, 2);
    });

    test('ten opens it', () {
      expect(unlocked(drum(tapsToUnlock)), isTrue);
    });

    test('nine does not', () {
      expect(unlocked(drum(tapsToUnlock - 1)), isFalse);
    });
  });

  group('the gap', () {
    /*
      A tap now and a tap tomorrow are two different intentions. Without the
      reset the count would creep up across ordinary use — nine taps over a
      week, then one more, and the card is open.
    */
    test('a pause starts over', () {
      final nearly = drum(9);
      final later = tap(nearly, at(9 * 200 + tapGap.inMilliseconds + 1));
      expect(later.count, 1);
      expect(unlocked(later), isFalse);
    });

    test('but the gap itself is not a pause', () {
      final one = tap(noTaps, at(0));
      expect(tap(one, at(tapGap.inMilliseconds)).count, 2);
    });

    test('and a fresh state is never mid-run', () {
      expect(noTaps.count, 0);
      expect(noTaps.last, isNull);
      expect(unlocked(noTaps), isFalse);
    });
  });

  group('the hint', () {
    /*
      Silence until the tapping is obviously deliberate, then count down.
      Starting the countdown at ten would announce the thing we just decided
      to hide.
    */
    test('nothing is said early on', () {
      expect(tapHint(drum(1)), isNull);
      expect(tapHint(drum(6)), isNull);
      expect(tapHint(drum(tapsToUnlock - 4)), isNull);
    });

    test('then it counts down', () {
      expect(tapHint(drum(tapsToUnlock - 3)), '3 more taps');
      expect(tapHint(drum(tapsToUnlock - 2)), '2 more taps');
      expect(tapHint(drum(tapsToUnlock - 1)), '1 more tap');
    });

    test('and stops once it is open', () {
      expect(tapHint(drum(tapsToUnlock)), isNull);
      expect(tapHint(drum(tapsToUnlock + 5)), isNull);
    });

    test('tapsLeft never goes negative', () {
      expect(tapsLeft(drum(tapsToUnlock + 5)), 0);
    });
  });

  group('staying unlocked', () {
    /*
      Once open it stays open until Hide. Testing a notification means leaving
      Settings, closing the app, waiting, and tapping the notification — and
      coming back to ten more taps every time turns a five-second check into a
      chore, which is how a test bench stops being used.
    */
    setUp(() => rememberUnlocked(false));
    tearDown(() => rememberUnlocked(false));

    test('locked to begin with', () => expect(readUnlocked(), isFalse));

    test('remembering opens it', () {
      rememberUnlocked(true);
      expect(readUnlocked(), isTrue);
    });

    test('and Hide closes it', () {
      rememberUnlocked(true);
      rememberUnlocked(false);
      expect(readUnlocked(), isFalse);
    });

    /*
      The web version needed sessionStorage for this, wrapped in try/catch
      because private browsing throws on storage access. A Flutter process is
      the session, so there is nothing to store and nothing that can throw —
      which is why there is no "storage refused, treat as locked" test here.
    */
  });
}
